### ISSUE TITLE
Test dependencies obscure/confusing and not DRY

### ISSUE THREAD

**Issue by ranger6** [fsaintjacques/semver-tool#74]
Looking at the GitHub action logs, I noticed the following:

```
asdf install semver latest
asdf global semver latest
semver --version
semver: 3.3.0
```
which seemed a bit strange because the action (unit tests and lint):

1. should have been running against a non-released commit of `semver`
2. appeared to be running against a non-latest release (we are at 3.4.0)

As well, the log output claim "asdf install semver latest" is simply the output of `echo "asdf install semver latest"`. The actual install specifies version 3.3.0.

After I ran a couple tests, it turns out that the GitHub action tests do run (AFAICS) against the appropriate `semver` commit. Ouff.

On the other hand, the output is confusing and it's not clear what the point of downloading a specific version and echoing the version string when that is not (in most cases) the code being tested.

And more: the GitHub action code requires/downloads version 1.3.0 of `Bats` while the project moved to version 1.6.0 a bit before the GitHub action code was committed.  On the upside, the action code did specify a version for the `spellcheck` dependency (the `Makefile` does not).

These problems seem to be the result of:

1. non-DRY parameterization
2. distributed dependency management
3. two separate testing paths: a `Makefile` running tests under Docker and using *nix install; a GitHub action script (ci.yaml) with it's own test execution scripting and asdf installation

Suggested steps to clean this up (in priority order):

1. At a minimum: specify exact and consistent versions for dependencies; maybe use variables rather than imbedded constants.
2. Remove or clarify/adjust the script that downloads a specific release of semver and prints its version string; then tests something else.
3. Create a little file with dependencies defined once and in one place.  Use these definitions in the various scripts and tools.
4. Eliminate the difference between local testing and automated (GitHub actions) testing: GitHub actions become wrappers around the local testing scripts/tools.  If `asdf` installation is to be tested (good idea! same for the other installation methods!), it should be explicitly tested: not by simply using it as part of the GitHub actions.

**Comment by ranger6** [comment]
I realized that I didn't broadcast this issue to those who might be interested.  So, here's the notification for @fsaintjacques and @mathew-fleisch !

**Comment by ranger6** [comment]
Mathew,

Well, this PR kind of clobbered your earlier contribution.  Just want to let you know that I believe the project moved forward due to that earlier contribution:

1.  You identified the lack of dependency spec. for shellcheck .. which is not unimportant.  It looks like the latest version will find problems with our scripts.  We’re set to cleanly move forward when the time comes.
2.  You cleaned up the CI script with the inclusion of PR’s and pushes.  I think Francois was a bit baffled at one point why the CI script was not running.
3.  The “asdf” plugin you built and your work to include it in the semver-tool project directed my attention to this interesting framework and helped clarify my thinking about how developers may want *their* tool versions and not the stable ones .. and how to do this explicitly and cleanly.  The project is better and I learned some stuff.
4.  And just by having an active community (e.g. your contribution and comments), we got the doc-test back into the test suite.  More eyeballs, more pointers for other’s eyeballs, just makes things better.

cheers,
— robert

**Comment by ranger6** [comment]
Closed with PR #75.

**Comment by mathew-fleisch** [comment]
No worries at all. You made great points and a great contribution.

I have been mainly a user of this project from the devops perspective. I manage many environments and systems that have a number of dependencies that need frequent updates to get security patches. For reasons™ companies will often pin dependencies to specific versions, so they don't have to deal with broken functionality. This semver project has been very useful at baking semver math/logic into CI pipelines as well as testing/setting specific versions. For instance, I maintain self-hosted github action runners, that run on my hybrid arm/x86 k3s cluster (raspberry pi 4b+ 8gb and an i5nuc) at home, to run all of my personal github action jobs. This runner is just a docker container, loaded with the github action agent and some build tools (installed with asdf). A cronjob, checks asdf for new versions of each of these dependencies, and uses the semver tool to bump-patch the container as a new version:
https://github.com/mathew-fleisch/github-actions-runner/blob/main/.github/workflows/auto-merge-dependabot-prs.yaml#L57

Great for setting/testing specific versions, AND in CI workflows. 😄
